platform/core/src/main/java/org/corfudb/platform/core/Result.java

```diff
@@ -0,0 +1,113 @@
+package org.corfudb.platform.core;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Container type for encapsulating function result and propagating errors.
+ *
+ * @author jameschang
+ * @since 2018-07-25
+ */
+public class Result<T, E extends RuntimeException> implements Supplier<T> {
+
+    private final T value;
+    private final E error;
+
+    public Result(T value) {
+        this(Objects.requireNonNull(value), null);
+    }
+
+    public Result(E error) {
+        this(null, Objects.requireNonNull(error));
+    }
+
+    private Result(T value, E error) {
+        // Exactly one value/error is non-null based on constructors calling this constructor.
+        this.value = value;
+        this.error = error;
+    }
+
+    /**
+     * Returns the encapsulated result if the {@link Result} instance is not an error.
+     *
+     * @return the encapsulated {@link #value}.
+     * @throws E if instance encapsulates an error.
+     */
+    @Override
+    public T get() {
+        if (error != null) {
+            throw error;
+        }
+
+        return value;
+    }
+
+    /**
+     * Returns the encapsulated error if the {@link Result} instance is an error.
+     *
+     * @return the encapsulated {@link #error}.
+     * @throws NoSuchElementException if there is no error present.
+     */
+    public E getError() {
```

Comment:
It would be better if there is another function to test whether there is any error by return true/false.